src/ssm-mem.c

```diff
@@ -130,30 +167,42 @@ void *ssm_mem_alloc(size_t size) {
   if (pool->free_list_head == END_OF_FREELIST)
     alloc_pool(p);
 
-  void *buf = pool->free_list_head->block_buf;
+  void *m = pool->free_list_head->block_buf;
 
-  if (pool->free_list_head->free_list_next == UNINITIALIZED_FREE_BLOCK)
-    pool->free_list_head += SSM_MEM_POOL_SIZE(p) / sizeof(block_t);
-  else
-    pool->free_list_head = pool->free_list_head->free_list_next;
+#ifndef NVALGRIND
+  // Tell Valgrind that we have allocated m as a chunk of pool.
+  //
+  // The memory range [m..m+size] is now considered DEFINED.
+  VALGRIND_MALLOCLIKE_BLOCK(m, size, 0, 1);
```

Comment:
This is how we tell Valgrind that we have allocated something.

This appears to also work even if the underlying page is malloc'ed (at least leak-checking still seems to work). Valgrind seems to get confused over memory permissions, but we manage that manually (using `VALGRIND_MAKE_MEM_` macros).